docs/recipes/UsageWithTypescript.md

```diff
@@ -6,259 +6,295 @@ hide_title: true
 
-And then do the same for the system reducer's slice of state:
+Since those are types, it's safe to export them directly from your store setup file such as `app/store.ts` and import them directly into other files.
 
-```ts
-// src/store/system/types.ts
+```ts title="app/store.ts"
+import { configureStore } from '@reduxjs/toolkit'
+// ...
 
-export interface SystemState {
-  loggedIn: boolean
-  session: string
-  userName: string
-}
+const store = configureStore({
+  reducer: {
+    one: oneSlice.reducer,
+    two: twoSlice.reducer
+  }
+})
+
+// highlight-start
+// Infer the `RootState` and `AppDispatch` types from the store itself
+export type RootState = ReturnType<typeof store.getState>
+export type AppDispatch = typeof store.dispatch
+// highlight-end
 ```
 
-Note that we are exporting these interfaces to reuse them later in reducers and action creators.
+### Define Typed Hooks
 
-### Type Checking Actions & Action Creators
+While it's possible to import the `RootState` and `AppDispatch` types into each component, it's better to create typed versions of the `useDispatch` and `useSelector` hooks for usage in your application. This is important for a couple reasons:
 
-We will be using string literals and using `typeof` to declare our action constants and infer types. Note that we are making a tradeoff here when we declare our types in a separate file. In exchange for separating our types into a separate file, we get to keep our other files more focused on their purpose. While this tradeoff can improve the maintainability of the codebase, it is perfectly fine to organize your project however you see fit.
+- For `useSelector`, it saves you the need to type `(state: RootState)` every time
+- For `useDispatch`, the default `Dispatch` type does not know about thunks or other middleware. In order to correctly dispatch thunks, you need to use the specific customized `AppDispatch` type from the store that includes the thunk middleware types, and use that with `useDispatch`. Adding a pre-typed `useDispatch` hook keeps you from forgetting to import `AppDispatch` where it's needed.
 
-Chat Action Constants & Shape:
+Since these are actual variables, not types, it's important to define them in a separate file such as `app/hooks.ts`, not the store setup file. This allows you to import them into any component file that needs to use the hooks, and avoids potential circular import dependency issues.
 
-```ts
-// src/store/chat/types.ts
-export const SEND_MESSAGE = 'SEND_MESSAGE'
-export const DELETE_MESSAGE = 'DELETE_MESSAGE'
+```ts title="app/hooks.ts"
+import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux'
+import { RootState, AppDispatch } from './store'
```

Comment:
Using 
`import type { RootState, AppDispatch } from './store'`
might help some bundlers that work on a per-file level without looking at the type of the imported token and otherwise introduce a runtime dependency to ./store